src/tree/identity_tree.rs

```diff
@@ -0,0 +1,652 @@
+                if val == Hash::ZERO {
+                    self.tree.set_leaf(leaf_idx as usize, Hash::ZERO);
+                } else {
+                    // We can expect here because the `reallocate` implementation for Vec<H::Hash> as DynamicTreeStorage does not fail
+                    self.tree.push(val).expect("Could not push leaf");
+                }
+            }
+        }
+
+        // Split off tree updates at the new root
+        // Since the root was already removed from the updates, we can use split_off to separate the updates non inclusive of the root
+        let current_tree_updates = self.tree_updates.split_off(root);
+
+        // Clean up any roots that are no longer needed
+        for root in self.tree_updates.keys() {
+            self.roots.remove(&root.hash);
+        }
+
+        self.tree_updates = current_tree_updates;
+    }
+
+    /// Construct an inclusion proof for a given leaf
+    /// If a root is provided, the proof is constructed from the specified root
+    /// Otherwise, the proof is constructed from the current canonical tree
+    pub fn inclusion_proof(
+        &self,
+        leaf: Hash,
+        root: Option<&Root>,
+    ) -> eyre::Result<Option<InclusionProof>> {
+        let leaf_idx = self.leaves.get(&leaf).ok_or_eyre("Leaf not found")?;
+
+        if let Some(root) = root {
+            if root.hash == self.tree.root() {
+                let proof = self.tree.proof(*leaf_idx as usize);
+                Ok(Some(InclusionProof::new(self.tree.root(), proof)))
+            } else {
+                let proof = self.construct_proof_from_root(*leaf_idx, root)?;
+                Ok(Some(InclusionProof::new(root.hash, proof)))
+            }
+        } else {
+            let proof = self.tree.proof(*leaf_idx as usize);
+            Ok(Some(InclusionProof::new(self.tree.root(), proof)))
+        }
+    }
+
+    /// Construct an inclusion proof for a given leaf at a specified root
+    pub fn construct_proof_from_root(
+        &self,
+        leaf_idx: u32,
+        root: &Root,
+    ) -> eyre::Result<Proof> {
+        // Get the updates at the specified root
+        let updates = self
+            .tree_updates
+            .get(root)
+            .ok_or_eyre("Could not find root in tree updates")?;
+
+        // Convert the leaf index to a storage index for easier indexing
+        let mut node_idx = leaf_to_storage_idx(leaf_idx, self.tree.depth());
+
+        let mut proof: Vec<Branch<Hash>> = vec![];
+
+        // Traverse the tree from the leaf to the root, constructing the proof along the way with precedence for the updated node values
+        while node_idx > 0 {
+            let sibling_idx = if node_idx % 2 == 0 {
+                node_idx - 1
+            } else {
+                node_idx + 1
+            };
+
+            // Check if the sibling is in the updates, otherwise get the node from the tree
```

Comment:
If `updates` is per a given `root` is there a change that we'll miss the node updates from the "previous root".

Let's say we have 0x03 on mainnet, 0x02 on polygon and 0x01 on optimism.
So in this case we'd have the dense memory tree on 0x01 and then 0x02 and 0x03 in `tree_updates`, right?

We fetch an inclusion proof using root 0x03, but a node that was updated in 0x02 is in the path of the inclusion proof. This code however only checks the updates for 0x03 so we miss that update from 0x02 and return an invalid inclusion proof.

Not sure if I'm reading this right, but we should definitely have a test case for this

Reply:
Ok, I was wrong here, I see now that `construct_storage_updates` flattens the previous updates.

https://github.com/worldcoin/world-tree/pull/71

Added a test